### ISSUE TITLE
RIP protocol support?

### ISSUE THREAD

**Issue by miekg** [google/metallb#9]
My poor router only does RIP. 

Adding RIP support might be a nice addition.

As I should brush off my Pis, this seems like a nice enhancement to work, esp, for teaching me some internals.

**Comment by danderson** [comment]
RIP support sgtm. I explicitly want VRRP and RIP support at some point.

If you want to work on this, the general plan would be to create a rip-speaker binary that behaves a lot like bgp-speaker... But speaks RIP, obviously :). Config will need some refactoring to have separate options for BGP and RIP, and so you can configure per address pool "I want these addresses to advertise on $protocol".

Alternatively... Maybe creating one speaker per protocol is wasteful, because of binary overhead. Maybe we should s/bgp-speaker/speaker/, and put all the protocols in the same binary. WDYT?

**Comment by danderson** [comment]
Assigning to you since you expressed interest. If it turns out you don't have time (which is fine!), lemme know and I'll put it on my todo pile.

**Comment by miekg** [comment]
So I now have this :) https://github.com/miekg/rip/

Not sure why you talk about binaries (but haven't looked closely add the code) - why is this not all compiled in?

**Comment by danderson** [comment]
\o/

As it happens I wrote a [hacking guide](docs/hacking.md) yesterday, and it has a description of the code layout. Currently, MetalLB is 2 binaries, one cluster-wide controller for IP assignment, and a per-node BGP speaker for advertising.

So, we should probably add the RIP support in the per-node BGP speaker (and rename it something more generic). One problem is that RIP uses bcast/mcast, so the pod must run with host networking instead of the sandboxed networking stack. BGP does not require this, so it's a slight privilege escalation if we built RIP into the same binary. It's probably not a big deal, it's just a minor config change in the manifests. And VRRP needs the same thing. WDYT?

**Comment by miekg** [comment]
Haven't read that far into the RIP spec :-)

Using the host's network stack for multicast sgtm; but can you actually do that with k8s? Or just make it privileged and then hack around?

**Comment by danderson** [comment]
Yeah, k8s supports that just fine, you just need to set `spec.hostNetwork: true` in the pod template, and the pod will get created with all the usual sandboxing, except the network stack is shared with the host. That's how various pieces of k8s like kube-proxy get scheduled with the power to mess with the host's netfilter stack.

**Comment by miekg** [comment]
Ah nice.

Anyway just reading RFC 2453 some more it seems you may get away without using multicast and just sending to a remote IP - once the whole thing works I need to check with my router if that actually is OK.

How did you test your BGP impl btw?

**Comment by danderson** [comment]
I used a couple of strategies. First I searched for BGP pcaps on google, and used wireshark to extract only the relevant message bytes (not the whole TCP/IP packet) into files in `internal/bgp/testdata`. One of the tests just tries to parse those.

I also added a "serdes" test, to check that deserialize(serialize(msg)) == msg. That doesn't prove you're speaking the _correct_ protocol, but at least it proves that you're consistently speaking _something_ :)

Finally, I used BIRD to do most of my interop testing. You can configure it to just speak the protocol without programming kernel routing tables, and then use `birdc` to inspect its internal state and see if it's happy about your peer.

It's also useful to keep a wireshark running while you're trying to talk to BIRD. Any mistakes you made in the protocol will show up as decoding errors in Wireshark, which is very handy.

Once you have an idea of what is acceptable to a RIP speaker, we should modify `test-bgp-router` to also listen for RIP, and change the minikube sandbox so that we can quickly iterate with both protocols. But to start, it's probably simpler to run BIRD by hand and point a test binary at it.

**Comment by miekg** [comment]
Using wireshark was a good tip - the wireformat looks good and wireshark can decode it. Current master branch of miekg/rip should be usuable.

(Now the bad news; my EA6900 router does either NAT or RIP it seems)

**Comment by miekg** [comment]
OK miekg/rip does RIP, but I'm going to abandon this approach because it doesn't work in my home network - I'll focus #12 next which should work regardless.

**Comment by danderson** [comment]
Okay, thanks for getting the protocol working! I'll pick this up later and test it against my unifi router, which apparently does speak RIP.

**Comment by danderson** [comment]
I'm pretty much never going to implement this myself, because I'm never going to use RIP. My ability to accept PRs right now is pretty crappy, but that's the only way this will ever get implemented :)
